File: scripts/upgrade-snyk-go-dependencies.py
#!/usr/bin/env python3
import argparse
import subprocess

import requests


def get_latest_commit_sha(name):
    url = f"https://api.github.com/repos/snyk/{name}/commits"
    headers = {
        "Accept": "application/vnd.github.v3+json",
    }
    response = requests.get(url, headers=headers)
    response.raise_for_status()
    commits = response.json()
    return commits[0]['sha']


def upgrade_go_mod(name, commit_sha):
    subprocess.run(['go', 'get', '-u', f'github.com/snyk/{name}@{commit_sha}'], cwd='./cliv2', check=True)
    subprocess.run(['go', 'mod', 'tidy'], cwd='./cliv2', check=True)


def upgrade_dep(name):
    commit_sha = get_latest_commit_sha(name)
    print(f"The most recent commit SHA for {name} is: {commit_sha}")
    upgrade_go_mod(name, commit_sha)


if __name__ == "__main__":
    try:
        parser = argparse.ArgumentParser(description="Download a Snyk CLI go mod dependency")
        parser.add_argument("--name", help="Repository name to download from (e.g., go-application-framework)",
                            default="", required=False)
        args = parser.parse_args()

        if args.name == "":
            upgrade_dep("go-application-framework")
            upgrade_dep("snyk-ls")
        else:
            upgrade_dep(args.name)
    except Exception as e:
        print(f"An error occurred: {e}")

File: scripts/upgrade-snyk-go-dependencies.go
package main

/**
 * This script is used to upgrade the go dependencies in the cliv2 project.
 *
 * It uses the GitHub API to fetch the latest commit SHA for the specified repository
 * and then uses the `go get` command to upgrade the dependency.
 *
 * Usage:
 * go run scripts/upgrade-snyk-go-dependencies.go --name=go-application-framework
 */
import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"os/exec"
	"regexp"
)

type Commit struct {
	SHA    string `json:"sha"`
	Commit struct {
		Message string `json:"message"`
	} `json:"commit"`
	Url string `json:"html_url"`
}

func isValidRepository(name string) bool {
	if name == "" {
		return false
	}

	match, err := regexp.MatchString("^[a-zA-Z0-9-]+$", name)
	if err != nil {
		log.Fatal("Not able to validate repo name", err)
	}
	return match
}

func getLatestCommitSHA(name string) (string, error) {
	if !isValidRepository(name) {
		return "", fmt.Errorf("Invalid repository name: %s", name)
	}
	url := fmt.Sprintf("https://api.github.com/repos/snyk/%s/commits", name)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Accept", "application/vnd.github.v3+json")

	client := http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}

	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("Failed to fetch commits. Status code: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	var commits []Commit
	if err := json.Unmarshal(body, &commits); err != nil {
		return "", err
	}

	if len(commits) == 0 {
		return "", fmt.Errorf("No commits found for repository %s", name)
	}

	fmt.Println("🔍 The most recent commit for", name)
	fmt.Println("Message: ", commits[0].Commit.Message)
	fmt.Println("Sha:     ", commits[0].SHA)
	fmt.Println("URL:     ", commits[0].Url)
	return commits[0].SHA, nil
}

func upgradeGoMod(name, commitSHA string) error {
	fmt.Println("🔍 Upgrading: ", name)
	fmt.Println("🔍 Commit SHA: ", commitSHA)

	cmd := exec.Command("go", "get", fmt.Sprintf("github.com/snyk/%s@%s", name, commitSHA))
	cmd.Dir = "./cliv2"
	if err := cmd.Run(); err != nil {
		return err
	}

	fmt.Println("🧹 Running go mod tidy...")
	cmd = exec.Command("go", "mod", "tidy")
	cmd.Dir = "./cliv2"
	if err := cmd.Run(); err != nil {
		return err
	}

	fmt.Println("🚀 Upgrade successful for:", name)
	return nil
}

func upgradeDep(name string, commit string) error {
	var err error

	if commit == "" {
		fmt.Println("🔍 No commit SHA provided, fetching latest commit...")
		commit, err = getLatestCommitSHA(name)
		if err != nil {
			return err
		}
	}

	if err = upgradeGoMod(name, commit); err != nil {
		return err
	}

	return nil
}

func main() {
	name := flag.String("name", "", "Repository name to download from (e.g., go-application-framework)")
	commit := flag.String("commit", "", "Commit SHA to upgrade to")
	flag.Parse()

	if *name == "" {
		if err := upgradeDep("go-application-framework", ""); err != nil {
			fmt.Printf("An error occurred: %v\n", err)
		}
		if err := upgradeDep("snyk-ls", ""); err != nil {
			fmt.Printf("An error occurred: %v\n", err)
		}
	} else {
		if err := upgradeDep(*name, *commit); err != nil {
			fmt.Printf("An error occurred: %v\n", err)
		}
	}
}
